File: algorithms/code/leetcode/lc137_single_number_ii/lc137_single_number_ii.py
# https://leetcode.com/problems/single-number-ii/
# #google

import unittest


class Solution:
    def singleNumber(self, nums: list[int]) -> int:
        counters = [0] * 32
        for num in nums:
            for i, _ in enumerate(counters):
                counters[i] += num & 1
                num >>= 1
        result = 0
        mask = 1
        for i, _ in enumerate(counters):
            if counters[i] % 3 != 0:
                result |= mask
            mask <<= 1
        return result


class TestCode(unittest.TestCase):
    def test_1112(self) -> None:
        nums = [1, 1, 1, 2]
        assert Solution().singleNumber(nums) == 2

    def test_4344533(self) -> None:
        nums = [4, 3, 4, 4, 5, 3, 3]
        assert Solution().singleNumber(nums) == 5

File: algorithms/code/leetcode/lc137_single_number_ii/lc137_single_number_ii.rs
// https://leetcode.com/problems/single-number-ii/
// #medium

pub struct Solution;

impl Solution {
    pub fn single_number(nums: Vec<i32>) -> i32 {
        let mut counters = vec![0; 32];
        for num in nums {
            let mut num = num;
            for counter in &mut counters {
                *counter += num & 1;
                num >>= 1;
            }
        }
        let mut result = 0;
        let mut mask = 1;
        for counter in counters {
            if counter % 3 != 0 {
                result |= mask;
            }
            mask <<= 1;
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_1112() {
        let nums = vec![1, 1, 1, 2];
        assert_eq!(Solution::single_number(nums), 2);
    }

    #[test]
    fn test_4344533() {
        let nums = vec![4, 3, 4, 4, 5, 3, 3];
        assert_eq!(Solution::single_number(nums), 5);
    }
}
